Source/EasyNetQ/Internals/AsyncLock.cs

```diff
@@ -10,32 +10,54 @@ namespace EasyNetQ.Internals
     ///     any release. You should only use it directly in your code with extreme caution and knowing that
     ///     doing so can result in application failures when updating to a new EasyNetQ release.
     /// </summary>
-    internal sealed class AsyncLock : IDisposable
+    public sealed class AsyncLock : IDisposable
     {
         private readonly SemaphoreSlim semaphore;
-        private readonly IDisposable semaphoreReleaser;
+        private readonly IDisposable releaser;
+        private readonly Task<IDisposable> releaserTask;
 
+        /// <summary>
+        ///     This is an internal API that supports the EasyNetQ infrastructure and not subject to
+        ///     the same compatibility as public APIs. It may be changed or removed without notice in
+        ///     any release. You should only use it directly in your code with extreme caution and knowing that
+        ///     doing so can result in application failures when updating to a new EasyNetQ release.
+        /// </summary>
         public AsyncLock()
         {
             semaphore = new SemaphoreSlim(1);
-            semaphoreReleaser = new SemaphoreSlimReleaser(semaphore);
+            releaser = new Releaser(semaphore);
+            releaserTask = Task.FromResult(releaser);
         }
 
-        public async Task<IDisposable> AcquireAsync(CancellationToken cancellationToken = default)
+        /// <summary>
+        /// Acquires a lock
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns>Releaser, which should be disposed to release a lock</returns>
+        public Task<IDisposable> AcquireAsync(CancellationToken cancellationToken = default)
         {
-            await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
-            return semaphoreReleaser;
+            var acquireTask = semaphore.WaitAsync(cancellationToken);
+            return acquireTask.IsCompleted
+                ? releaserTask
+                : acquireTask.ContinueWith(
+                    (_, state) => (IDisposable)state,
+                    releaser,
+                    default,
+                    TaskContinuationOptions.ExecuteSynchronously,
+                    TaskScheduler.Default
+                );
         }
 
-        private class SemaphoreSlimReleaser : IDisposable
+        private sealed class Releaser : IDisposable
```

Comment:
I think this variant will more closely match the intentions and dispose pattern.

```c# 
private sealed class Releaser : IDisposable
        {
            private readonly SemaphoreSlim semaphore;
            private bool released;

            public Releaser(SemaphoreSlim semaphore) => this.semaphore = semaphore;

            public void Dispose()
            {
               if (!released)
               {
                    released = true;
                    semaphore.Release();
               }
        }
```

Reply:
Forget it, `Releaser` should be stateless.